File: src/main/java/arcana/common/research/requirements/SpecialRequirement.java
package arcana.common.research.requirements;

import arcana.common.capability.Knowledge;
import arcana.common.research.Requirement;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.util.ResourceLocation;

public class SpecialRequirement implements Requirement {
    ResourceLocation id;

    public SpecialRequirement() {}

    @Override
    public boolean satisfied(PlayerEntity player) {
        return player.getCapability(Knowledge.CAPABILITY).orElse(null).specialRequirementsMet.contains(id);
    }

    @Override
    public void take(PlayerEntity player) {
    }

    @Override
    public void deserialize(String data) {
        id = new ResourceLocation(data);
    }
}

File: src/main/kotlin/arcana/common/research/requirements/SpecialRequirement.kt
package arcana.common.research.requirements

import arcana.common.capability.getKnowledge
import arcana.common.research.Requirement
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.util.ResourceLocation

class SpecialRequirement : Requirement {
    var id: ResourceLocation? = null
    override fun satisfied(player: PlayerEntity): Boolean {
        return player.getKnowledge().specialRequirementsMet.contains(id)
    }

    override fun take(player: PlayerEntity) {}
    override fun deserialize(data: String) {
        id = ResourceLocation(data)
    }
}